docs/madr/decisions/011-mesh-traffic-route.md

```diff
@@ -0,0 +1,902 @@
+         - weight: 100
+           kind: MeshServiceSubset
+           name: backend
+           tags:
+            version: v2
+---
+metadata:
+ name: consumer
+spec:
+ targetRef:
+  kind: MeshService
+  name: frontend
+ to:
+  - targetRef:
+     kind: MeshService
+     name: backend
+    default:
+     rules:
+      - matches:
+         path:
+          prefix: /v2
+        filters:
+         - requestHeaderModifier:
+            add:
+             - name: env
+               value: dev
+        backendRefs:
+         - weight: 100
+           kind: MeshServiceSubset
+           name: backend
+           tags:
+            version: v2
+```
+
+The rules for merging a field in a `targetRef` policy is that more specific
+routes replace the value from less specific routes. So the final value of `rules`
+is the value from the `consumer` route:
+
+```yaml
+spec:
+ targetRef:
+  kind: MeshService
+  name: frontend
+ to:
+  - targetRef:
+     kind: MeshService
+     name: backend
+    rules:
+     - matches:
+        path:
+         prefix: /v2
+       filters:
+        - requestHeaderModifier:
+           add:
+            - name: env
+              value: dev
+       backendRefs:
+        - weight: 100
+          kind: MeshServiceSubset
+          name: backend
+          tags:
+           version: v2
+```
+
+This is likely surprising, especially because `consumer` didn't specify anything
+at all for `/v1`.
+
+##### Solution
+
+This MADR proposes to distribute `default` down into individual `rules` and merge
+`rules` based on _structural equality_ of the `matches` value.
+So we would instead have:
+
+```yaml
+metadata:
+ name: owner
+spec:
+ targetRef:
+  kind: Mesh
+ to:
+  - targetRef:
+     kind: MeshService
+     name: backend
+    rules:
+     - matches:
+        path:
+         prefix: /v1
+       default:
+        backendRefs:
+         - weight: 100
+           kind: MeshServiceSubset
+           name: backend
+           tags:
+            version: v1
+     - matches:
+        path:
+         prefix: /v2
+       default:
+        backendRefs:
+         - weight: 100
+           kind: MeshServiceSubset
+           name: backend
+           tags:
+            version: v2
+---
+metadata:
+ name: consumer
+spec:
+ targetRef:
+  kind: MeshService
+  name: frontend
+ to:
+  - targetRef:
+     kind: MeshService
+     name: backend
+    rules:
+     - matches:
+        path:
+         prefix: /v2
+       default:
+        filters:
+         - responseHeaderModifier:
+            add:
+             - name: env
+               value: dev
+```
+
+which gives us the final list of rules:
+
+```yaml
+spec:
+ targetRef:
+  kind: MeshService
+  name: frontend
+ to:
+  - targetRef:
+     kind: MeshService
+     name: backend
+    rules:
+      - matches:
+         path:
+          prefix: /v1
+        backendRefs:
+         - weight: 100
+           kind: MeshServiceSubset
+           name: backend
+           tags:
+            version: v1
+     - matches:
```

Comment:
How are we going to define an equivalence relation for `matches`? Are these matches equal:
```
matches:
  path: 
    prefix: /v1
```
and
```
matches:
  path: 
    exact: /v1
```

Reply:
What if `owner` changes the `matches` from `prefix` to `exact`. Now `consumer` instead of overriding the existing route creates a new route.

Reply:
I think it's ok behaviour to stick with exact equality of matches.

Reply:
I don't really like the idea of using `matches` as keys. I'd rather add a `name` field for each rule and use it as a key

Reply:
Doesn't that potentially create kind of a dual issue where by referencing with name, now the `consumer` adds something for the `prefix` when they wanted just something for `exact`?

I think the idea of having the merging is more to give `owner` a way to avoid having their routes overwritten than to give `consumer` a way to ensure that `owner`s routes are applied. I.e. why should `consumer` care if they've created a new route?

Reply:
I agree, but even if `owner` is motivated to have their routes in the `consumer` proxy, I think it's still difficult for `owner` to check if routes were overwritten or not. As `owner` you can't quickly grep all MeshHTTPRoutes in the mesh and check references on your routes.

Reply:
Shouldn't you just be using `override` (not implemented yet) if you want to make sure it's not overwritten?

I think the real use case here is simply have some routes defined in one policy and still have these work for other policies (think standard inspection API for all the apps on one side, and core api on the other).

I think the whole use case of customising a route defined somewhere else doesn't hold much real world use case. Though I'm happy to be shown use-cases

Reply:
Image if the value of `matches` is big and contains lists of `headers` and `queryParams`. It's going to be difficult for `customer` to keep these things in sync

Reply:
To be clear, I don't think the `name`-based merging helps in this case, does it?

We talked about giving semantics to `matches` and it's probably a workable idea? But might be very complex to specify..

Reply:
I also don't think `name`-based merging is going to help in this case. I just brought it here since this thread is sort of about "matches as a key" problem

Reply:
Isn't the first problem here that the `customer` route with header match won't have the `owner`'s rules? So even without deactivating any `owner` routes or changing any names, the new `customer` route already "breaks" the _`owner`'s_ routes because they're no longer applied if the header matches?

Reply:
> They may very well want to add additional filters. I.e. the customer and owner routes could compose rather than overwrite (why I wanted a filter merging functionality).

In that case, we can create an `HTTPFilter` policy. This policy applies modification filters on existing routes. So `MeshTimeout`, `MeshRetries`, and `HTTPFilter` policies are used to _alter_ existing routes.  And `MeshHTTPRoute` (or shall we rename it to `MeshTrafficControl`?) is used to _define_ routes. If you override the policy that _defines_ routes, it shouldn't be a surprise that other routes are gone. And also it's not surprising because our behavior is consistent across different policies, we say "arrays under `default` section always replace each other".

Reply:
> And also it's not surprising because our behavior is consistent across different policies, we say "arrays under default section always replace each other".

Sure, given the spec having `default.rules`, if the user is the ideal perfect Kuma user and has this rule in they're head first and foremost, it's more surprising in the sense that we would have given the user a useless policy and a useless spec with which to write composable routes. But maybe they have something like gateway API in their head, where multiple route resources are appended together? And maybe they assume our routes function like those routes, in which case it's very easy to assume, or want that, a list of rules works the same way, in spite of it appearing under `default`.

Anyway the rule still applies, we just put `default` somewhere else than the top-level.

I think that a user is more likely to assume that `(filters, backendRefs)` for disjoint `matches` are handled independently from each other as opposed to as lists that overwrite each other. That this is a more useful way to express routing rules. That they expect rather a `map[Matches]FiltersAndBackendRefs` at the top level under `default` rather than `[]MatchesFiltersAndBackendRefs`. And that`Matches` has a natural ordering to make it behave even more as expected with respect to "submatches" like we talked about above.

> In that case, we can create an HTTPFilter policy. This policy applies modification filters on existing routes. So MeshTimeout, MeshRetries, and HTTPFilter policies are used to alter existing routes. And MeshHTTPRoute (or shall we rename it to MeshTrafficControl?) is used to define routes.

Does this suggestion come from the axiom "everything must follow policy attachment with `default`/`override` at the top level"? I don't see why that needs to be true. It's just a rule for a generic form of policy. There's no reason every resource needs to look like this.

I'm not sure how this would look, an example would help. I suspect it's exactly as powerful as putting the filters into the `MeshHTTPRoute` resource, though.

Reply:
To me the user stories are:

1. Define a different set of routes in different policies (one resource at `kind:Mesh` and another at `kind:MeshService`). In this case it's ok to consider these as non overlapping.
2. A user want to override filters or backendRefs of an inherited rule. In that case being of `matches` as key if the parent changes the matches then the override won't happen. I feel like even though this is real I'm not sure people really change `matches` after setting them.